src/main/java/org/jitsi/jicofo/AbstractChannelAllocator.java

```diff
@@ -0,0 +1,393 @@
+     * First argument stands for "start audio muted" and the second one for
+     * "start video muted". The information is included as a custom extension in
+     * 'session-initiate' sent to the user.
+     */
+    protected final boolean[] startMuted;
+
+    /**
+     * Indicates whether or not this thread will be doing a "re-invite". It
+     * means that we're going to replace previous conference which has failed.
+     * Channels are allocated on new JVB and peer is re-invited with
+     * 'transport-replace' Jingle action as opposed to 'session-initiate' in
+     * regular invite.
+     */
+    protected final boolean reInvite;
+
+    /**
+     * Initializes a new {@link AbstractChannelAllocator} instance which is to
+     * invite a specific {@link Participant} into a specific
+     * {@link JitsiMeetConferenceImpl} (using a specific jitsi-videobridge
+     * instance specified by a
+     * {@link org.jitsi.jicofo.JitsiMeetConferenceImpl.BridgeSession}).
+     *
+     * @param meetConference the {@link JitsiMeetConferenceImpl} into which to
+     * invite {@code participant}.
+     * @param bridgeSession the
+     * {@link org.jitsi.jicofo.JitsiMeetConferenceImpl.BridgeSession} which
+     * identifies the jitsi-videobridge instance on which to allocate channels.
+     * @param participant the participant to be invited.
+     * @param startMuted an array which must have the size of 2 where the first
+     * value stands for "start audio muted" and the second one for "video
+     * muted". This is to be included in client's offer.
+     * @param reInvite whether to send an initial offer (session-initiate) or
+     * a an updated offer (transport-replace).
+     */
+    protected AbstractChannelAllocator(
+            JitsiMeetConferenceImpl meetConference,
+            JitsiMeetConferenceImpl.BridgeSession bridgeSession,
+            AbstractParticipant participant,
+            boolean[] startMuted,
+            boolean reInvite)
+    {
+        this.meetConference = meetConference;
+        this.bridgeSession = bridgeSession;
+        this.participant = participant;
+        this.startMuted = startMuted;
+        this.reInvite = reInvite;
+        this.logger = Logger.getLogger(classLogger, meetConference.getLogger());
+    }
+
+    /**
+     * Entry point for the {@link AbstractChannelAllocator} task.
+     */
+    @Override
+    public void run()
+    {
+        try
+        {
+            doRun();
+        }
+        catch (Throwable e)
+        {
+            logger.error("Channel allocator failed: ", e);
+        }
+        finally
+        {
+            if (participant != null)
+            {
+                participant.channelAllocatorCompleted(this);
+            }
+        }
+    }
+
+    private void doRun()
+        throws OperationFailedException
+    {
+        List<ContentPacketExtension> offer;
+
+        if (canceled)
+        {
+            return;
+        }
+
+        offer = createOffer();
+        if (canceled)
```

Comment:
The repeated checks for canceled are because some of the calls wait for network operations (e.g. allocateCandidates). I'll use "task" instead of "thread"

Reply:
There's no really an easy way to kill/shutdown a thread in java and as far as I know it was recommended to use a flag. Although maybe it could be done better or in a cleaner way than I did. If I remember correctly the intention there was to check for cancel just before/after some long running operation like sending XMPP message.

Reply:
Yes, I think I added more checks for 'canceled' just to be on the safe side, and they might not make sense.